nix/unflatten-drv.nix

```diff
@@ -0,0 +1,34 @@
+# Unflattens a derivation. If a derivation produces a single file instead of a
+# directory, this derivation symlinks the artifact to a result directory.
+
+{
+  runCommandLocal,
+}:
+
+{
+  # Derivation to flatten.
+  drv,
+  # Artifact path inside the derivation.
+  artifactPath ? drv.name,
+  # Name of the new derivation.
+  name ? "unflattened-${drv.name}",
+}:
+
+runCommandLocal name
+  {
+    passthru = drv.passthru;
+  }
+  ''
+    set -euo pipefail
+
+    mkdir -p $out/$(dirname ${artifactPath})
+
+    if [ -e "${drv}" ] && { [ -f "${drv}" ] || [ -L "${drv}" ]; }
```

Comment:
Why the explicit existence check? Both `-f` and `-L` imply `-e`. This does not say anything about the validity of the symlink target.